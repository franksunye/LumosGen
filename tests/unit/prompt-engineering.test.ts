/**
 * ÊèêÁ§∫Â∑•Á®ãÁ≥ªÁªü‰∏ìÈ°πÊµãËØï - VitestÁâàÊú¨
 * È™åËØÅÊèêÁ§∫Ê®°Êùø„ÄÅ‰∏ä‰∏ãÊñáÊ≥®ÂÖ•„ÄÅË¥®ÈáèÊéßÂà∂ÁöÑÊ†∏ÂøÉÂäüËÉΩ
 * ‰ªéËá™ÂÆö‰πâÊµãËØïÊ°ÜÊû∂ËøÅÁßªÂà∞Vitest
 */

import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest'

// Mock VS Code API
vi.mock('vscode', () => ({
  workspace: {
    workspaceFolders: [{
      uri: { fsPath: '/test/workspace' }
    }]
  },
  window: {
    createOutputChannel: vi.fn(() => ({
      appendLine: vi.fn(),
      show: vi.fn(),
      clear: vi.fn()
    }))
  }
}))

// ÊµãËØïÈ°πÁõÆÂàÜÊûêÊï∞ÊçÆ
const mockProjectAnalysis = {
  metadata: {
    name: 'TestProject',
    description: 'A revolutionary AI-powered development platform',
    version: '1.0.0',
    author: 'Test Developer',
    license: 'MIT'
  },
  techStack: ['TypeScript', 'React', 'Node.js', 'Express', 'MongoDB'],
  features: [
    'AI Content Generation',
    'Real-time Processing',
    'Advanced Analytics',
    'User-friendly Interface',
    'Enterprise Security',
    'Multi-platform Support'
  ],
  useCases: [
    'AI-powered development',
    'Content automation',
    'Real-time analytics',
    'Enterprise solutions'
  ],
  targetAudience: 'Developers and enterprises',
  marketingGoals: [
    'Increase developer adoption',
    'Showcase AI capabilities',
    'Highlight ease of use',
    'Demonstrate ROI'
  ]
}

// Mock AIÊúçÂä°
class MockAIService {
  async generateContent(prompt: string, options: any = {}) {
    // Ê®°ÊãüAIÂìçÂ∫îÔºåÂü∫‰∫éÊèêÁ§∫ËØçÂÜÖÂÆπÁîüÊàêÁõ∏Â∫îÁöÑÂõûÂ§ç
    const response = {
      content: this.generateMockResponse(prompt),
      tokens: Math.floor(prompt.length / 4),
      cost: 0.001,
      provider: 'mock',
      model: 'mock-model'
    }
    
    // Ê®°ÊãüÂìçÂ∫îÂª∂Ëøü
    await new Promise(resolve => setTimeout(resolve, 100))
    return response
  }
  
  generateMockResponse(prompt: string): string {
    if (prompt.includes('homepage') || prompt.includes('Homepage')) {
      return `# Welcome to TestProject

## Revolutionary AI-Powered Solution

TestProject is a cutting-edge TypeScript and React application that transforms how developers work with AI integration and real-time processing.

### Key Features
- ü§ñ Advanced AI Integration
- ‚ö° Real-time Processing
- üéØ User-friendly Interface

### Quick Start
\`\`\`bash
npm install test-project
npm start
\`\`\`

Ready to revolutionize your workflow? Get started today!`
    } else if (prompt.includes('about') || prompt.includes('About')) {
      return `## About TestProject

TestProject represents the next generation of AI-powered development tools. Built with TypeScript and React, our platform combines cutting-edge artificial intelligence with intuitive user experience.

### Our Mission
To democratize AI development and make advanced machine learning accessible to every developer.

### Technology Stack
- Frontend: React with TypeScript
- Backend: Node.js with Express
- Database: MongoDB
- AI: Advanced machine learning models

### Why Choose TestProject?
1. **Ease of Use**: Intuitive interface designed for developers
2. **Powerful AI**: State-of-the-art machine learning capabilities
3. **Scalable**: Enterprise-ready architecture
4. **Community**: Active developer community and support`
    } else if (prompt.includes('blog') || prompt.includes('Blog') || prompt.includes('Getting Started') || prompt.includes('Write a blog post')) {
      return `# Getting Started with TestProject: A Developer's Guide

Are you ready to supercharge your development workflow with AI? TestProject makes it easier than ever to integrate powerful artificial intelligence into your applications.

## What Makes TestProject Special?

TestProject isn't just another development tool ‚Äì it's a complete AI-powered platform that understands your needs as a developer.

### Key Benefits:
- **Rapid Development**: Cut development time by 50%
- **AI Integration**: Seamless AI capabilities out of the box
- **TypeScript Support**: Full type safety and IntelliSense
- **React Components**: Pre-built UI components

## Getting Started

Installation is simple:

\`\`\`bash
npm install test-project
\`\`\`

Then import and use:

\`\`\`typescript
import { TestProject } from 'test-project';

const app = new TestProject({
  aiEnabled: true,
  realTimeProcessing: true
});
\`\`\`

## Conclusion

TestProject is more than a tool ‚Äì it's your AI development partner. Start building the future today!`
    }
    
    return `Generated content based on prompt: ${prompt.substring(0, 100)}...`
  }
}

// MockÊèêÁ§∫Ê®°ÊùøÂ∫ì
class MockPromptTemplateLibrary {
  private templates: Map<string, any> = new Map()
  private aiService: MockAIService

  constructor() {
    this.aiService = new MockAIService()
    this.initializeTemplates()
  }

  private initializeTemplates() {
    this.templates.set('homepage', {
      name: 'Homepage Template',
      description: 'Generate compelling homepage content for projects',
      structure: ['hero', 'features', 'quickstart', 'cta'],
      validationRules: ['min_length', 'includes_features', 'has_cta'],
      template: `Create a compelling homepage for {{PROJECT_NAME}}.

Project: {{PROJECT_NAME}}
Description: {{DESCRIPTION}}
Tech Stack: {{TECH_STACK}}
Key Features: {{FEATURES}}
Target Audience: {{AUDIENCE}}
Tone: {{TONE}}

Generate content that is {{TONE}} and appeals to {{AUDIENCE}}.
{{OPTIMIZATION_STRATEGY}}

Include:
- Engaging hero section
- Feature highlights
- Quick start guide
- Clear call-to-action`
    })

    this.templates.set('about', {
      name: 'About Page Template',
      description: 'Generate comprehensive about page content',
      structure: ['mission', 'technology', 'benefits', 'team'],
      validationRules: ['min_length', 'includes_mission', 'mentions_tech'],
      template: `Create an about page for {{PROJECT_NAME}}.

Project: {{PROJECT_NAME}}
Description: {{DESCRIPTION}}
Mission: {{MISSION}}
Technology: {{TECH_STACK}}
Tone: {{TONE}}

Generate {{TONE}} content that explains:
- Our mission and vision
- Technology stack and approach
- Benefits for users
- Why choose this project
{{OPTIMIZATION_STRATEGY}}`
    })

    this.templates.set('blog', {
      name: 'Blog Post Template',
      description: 'Generate engaging blog post content',
      structure: ['introduction', 'main_content', 'examples', 'conclusion'],
      validationRules: ['min_length', 'has_examples', 'engaging_intro'],
      template: `Write a blog post about {{PROJECT_NAME}}.

Topic: Getting Started with {{PROJECT_NAME}}
Project: {{PROJECT_NAME}}
Description: {{DESCRIPTION}}
Features: {{FEATURES}}
Target Audience: {{AUDIENCE}}
Tone: {{TONE}}

Create a {{TONE}} blog post that:
- Introduces the project compellingly
- Explains key benefits
- Provides practical examples
- Includes code snippets
- Ends with a strong call-to-action
{{OPTIMIZATION_STRATEGY}}`
    })
  }

  getAvailableTemplates(): string[] {
    return Array.from(this.templates.keys())
  }

  getTemplateInfo(templateName: string) {
    return this.templates.get(templateName)
  }

  generatePrompt(templateName: string, projectAnalysis: typeof mockProjectAnalysis, options: any = {}): string {
    const template = this.templates.get(templateName)
    if (!template) {
      throw new Error(`Template not found: ${templateName}`)
    }

    let prompt = template.template

    // ÊõøÊç¢Ê®°ÊùøÂèòÈáè
    const variables = {
      '{{PROJECT_NAME}}': projectAnalysis.metadata?.name || 'Unknown Project',
      '{{DESCRIPTION}}': projectAnalysis.metadata?.description || 'A great project',
      '{{TECH_STACK}}': (projectAnalysis.techStack || []).join(', ') || 'Modern technologies',
      '{{FEATURES}}': (projectAnalysis.features || []).join(', ') || 'Great features',
      '{{AUDIENCE}}': options.audience || projectAnalysis.targetAudience || 'users',
      '{{TONE}}': options.tone || 'professional',
      '{{MISSION}}': options.mission || 'To provide innovative solutions',
      '{{OPTIMIZATION_STRATEGY}}': this.getOptimizationStrategy(options.optimizationStrategy)
    }

    for (const [variable, value] of Object.entries(variables)) {
      prompt = prompt.replace(new RegExp(variable, 'g'), value)
    }

    return prompt
  }

  private getOptimizationStrategy(strategy?: string): string {
    switch (strategy) {
      case 'concise':
        return '\nKeep the content concise and to the point.'
      case 'detailed':
        return '\nProvide detailed explanations and comprehensive information.'
      case 'creative':
        return '\nUse creative language and engaging storytelling.'
      default:
        return '\nBalance clarity with engagement.'
    }
  }

  async generateContent(templateName: string, projectAnalysis: typeof mockProjectAnalysis, options: any = {}) {
    const prompt = this.generatePrompt(templateName, projectAnalysis, options)
    return await this.aiService.generateContent(prompt, options)
  }

  validatePrompt(prompt: string, templateName: string): { isValid: boolean; errors: string[]; score: number } {
    const errors: string[] = []
    let score = 100

    // Âü∫Êú¨È™åËØÅ
    if (!prompt || prompt.trim().length === 0) {
      errors.push('Prompt cannot be empty')
      score -= 50
    }

    if (prompt.length < 100) {
      errors.push('Prompt is too short (minimum 100 characters)')
      score -= 20
    }

    if (prompt.length > 10000) {
      errors.push('Prompt is too long (maximum 10000 characters)')
      score -= 10
    }

    // Ê®°ÊùøÁâπÂÆöÈ™åËØÅ
    const template = this.templates.get(templateName)
    if (template) {
      for (const rule of template.validationRules) {
        switch (rule) {
          case 'min_length':
            if (prompt.length < 200) {
              errors.push('Content does not meet minimum length requirement')
              score -= 15
            }
            break
          case 'includes_features':
            if (!prompt.includes('feature') && !prompt.includes('Feature')) {
              errors.push('Content should mention features')
              score -= 10
            }
            break
          case 'has_cta':
            if (!prompt.includes('start') && !prompt.includes('get') && !prompt.includes('try')) {
              errors.push('Content should include a call-to-action')
              score -= 10
            }
            break
        }
      }
    }

    return {
      isValid: errors.length === 0,
      errors,
      score: Math.max(0, score)
    }
  }

  optimizePrompt(prompt: string, strategy: string = 'balanced'): string {
    switch (strategy) {
      case 'concise':
        return prompt.replace(/\n\n+/g, '\n').substring(0, Math.floor(prompt.length * 0.8))
      case 'detailed':
        return prompt + '\n\nProvide additional context and detailed explanations where appropriate.'
      case 'creative':
        return prompt.replace(/Generate/, 'Creatively craft').replace(/Create/, 'Imaginatively design')
      default:
        return prompt
    }
  }
}

describe('Prompt Engineering Unit Tests', () => {
  let promptLibrary: MockPromptTemplateLibrary
  let aiService: MockAIService
  let testOptions: any

  beforeEach(() => {
    console.log('üîß Setting up prompt engineering tests...')
    promptLibrary = new MockPromptTemplateLibrary()
    aiService = new MockAIService()
    testOptions = {
      tone: 'professional',
      audience: 'developers',
      optimizationStrategy: 'balanced',
      includeExamples: true,
      maxLength: 2000
    }
  })

  afterEach(() => {
    // Ê∏ÖÁêÜÊµãËØïÁéØÂ¢É
  })

  describe('ÊèêÁ§∫Ê®°ÊùøÁªìÊûÑ', () => {
    it('Â∫îËØ•Êèê‰æõÂü∫Êú¨Ê®°Êùø', () => {
      const templates = promptLibrary.getAvailableTemplates()

      expect(templates.length).toBeGreaterThanOrEqual(3)
      expect(templates).toContain('homepage')
      expect(templates).toContain('about')
      expect(templates).toContain('blog')
    })

    it('Â∫îËØ•È™åËØÅÊ®°ÊùøÁªìÊûÑÂÆåÊï¥ÊÄß', () => {
      const templates = promptLibrary.getAvailableTemplates()

      for (const templateName of templates) {
        const templateInfo = promptLibrary.getTemplateInfo(templateName)

        expect(templateInfo.name).toBeTruthy()
        expect(templateInfo.name.toLowerCase()).toContain(templateName)
        expect(templateInfo.description.length).toBeGreaterThan(10)
        expect(Array.isArray(templateInfo.structure)).toBe(true)
        expect(templateInfo.structure.length).toBeGreaterThan(0)
        expect(Array.isArray(templateInfo.validationRules)).toBe(true)
        expect(templateInfo.validationRules.length).toBeGreaterThan(0)
      }
    })

    it('Â∫îËØ•ÂåÖÂê´ÂøÖË¶ÅÁöÑÊ®°ÊùøÂÖÉÁ¥†', () => {
      const homepageTemplate = promptLibrary.getTemplateInfo('homepage')

      expect(homepageTemplate.structure).toContain('hero')
      expect(homepageTemplate.structure).toContain('features')
      expect(homepageTemplate.structure).toContain('cta')
      expect(homepageTemplate.validationRules).toContain('min_length')
      expect(homepageTemplate.validationRules).toContain('includes_features')
    })
  })

  describe('‰∏ä‰∏ãÊñáÊ≥®ÂÖ•', () => {
    it('Â∫îËØ•Ê≠£Á°ÆÊ≥®ÂÖ•È°πÁõÆ‰ø°ÊÅØ', () => {
      const prompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)

      expect(prompt).toContain(mockProjectAnalysis.metadata.name)
      expect(prompt).toContain(mockProjectAnalysis.metadata.description)
      expect(prompt).toContain('TypeScript')
      expect(prompt).toContain('React')
      expect(prompt).toContain('AI Content Generation')
    })

    it('Â∫îËØ•Ê≥®ÂÖ•ÈÄâÈ°πÈÖçÁΩÆ', () => {
      const prompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)

      expect(prompt).toContain(testOptions.tone)
      expect(prompt).toContain(testOptions.audience)
    })

    it('Â∫îËØ•Â§ÑÁêÜÁº∫Â§±ÁöÑÈÄâÈ°π', () => {
      const prompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, {})

      expect(prompt).toBeTruthy()
      expect(prompt).toContain(mockProjectAnalysis.metadata.name)
      expect(prompt).toContain('professional') // ÈªòËÆ§tone
    })

    it('Â∫îËØ•‰∏∫‰∏çÂêåÊ®°ÊùøÊ≥®ÂÖ•ÈÄÇÂΩìÁöÑ‰∏ä‰∏ãÊñá', () => {
      const homepagePrompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)
      const aboutPrompt = promptLibrary.generatePrompt('about', mockProjectAnalysis, testOptions)
      const blogPrompt = promptLibrary.generatePrompt('blog', mockProjectAnalysis, testOptions)

      expect(homepagePrompt).toContain('hero section')
      expect(aboutPrompt).toContain('mission')
      expect(blogPrompt).toContain('blog post')

      // ÊâÄÊúâÊ®°ÊùøÈÉΩÂ∫îËØ•ÂåÖÂê´È°πÁõÆÂêçÁß∞
      expect(homepagePrompt).toContain(mockProjectAnalysis.metadata.name)
      expect(aboutPrompt).toContain(mockProjectAnalysis.metadata.name)
      expect(blogPrompt).toContain(mockProjectAnalysis.metadata.name)
    })
  })

  describe('ÊèêÁ§∫‰ºòÂåñÁ≠ñÁï•', () => {
    it('Â∫îËØ•ÊîØÊåÅ‰∏çÂêåÁöÑ‰ºòÂåñÁ≠ñÁï•', () => {
      const strategies = ['concise', 'detailed', 'creative']

      for (const strategy of strategies) {
        const options = { ...testOptions, optimizationStrategy: strategy }
        const prompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, options)

        expect(prompt.length).toBeGreaterThan(100)

        if (strategy === 'concise') {
          expect(prompt).toContain('concise')
        } else if (strategy === 'detailed') {
          expect(prompt).toContain('detailed')
        } else if (strategy === 'creative') {
          expect(prompt).toContain('creative')
        }
      }
    })

    it('Â∫îËØ•‰ºòÂåñÁé∞ÊúâÊèêÁ§∫', () => {
      const basePrompt = 'Generate content for TestProject. Create engaging material.'

      const concisePrompt = promptLibrary.optimizePrompt(basePrompt, 'concise')
      const detailedPrompt = promptLibrary.optimizePrompt(basePrompt, 'detailed')
      const creativePrompt = promptLibrary.optimizePrompt(basePrompt, 'creative')

      expect(concisePrompt.length).toBeLessThanOrEqual(basePrompt.length)
      expect(detailedPrompt.length).toBeGreaterThan(basePrompt.length)
      expect(creativePrompt).toContain('Creatively craft')
      expect(creativePrompt).toContain('Imaginatively design')
    })
  })

  describe('Á´ØÂà∞Á´ØÊèêÁ§∫ÁîüÊàê', () => {
    it('Â∫îËØ•ÁîüÊàêÂÆåÊï¥ÁöÑAIÂìçÂ∫î', async () => {
      const result = await promptLibrary.generateContent('homepage', mockProjectAnalysis, testOptions)

      expect(result.content).toBeTruthy()
      expect(result.content.length).toBeGreaterThan(100)
      expect(result.tokens).toBeGreaterThan(0)
      expect(result.cost).toBeGreaterThan(0)
      expect(result.provider).toBe('mock')

      // È™åËØÅÂÜÖÂÆπË¥®Èáè
      expect(result.content).toContain('TestProject')
      expect(result.content).toContain('AI')
      expect(result.content).toContain('TypeScript')
    })

    it('Â∫îËØ•‰∏∫‰∏çÂêåÊ®°ÊùøÁîüÊàê‰∏çÂêåÂÜÖÂÆπ', async () => {
      const homepageResult = await promptLibrary.generateContent('homepage', mockProjectAnalysis, testOptions)
      const aboutResult = await promptLibrary.generateContent('about', mockProjectAnalysis, testOptions)
      const blogResult = await promptLibrary.generateContent('blog', mockProjectAnalysis, testOptions)

      // È™åËØÅÊâÄÊúâÂÜÖÂÆπÈÉΩÂåÖÂê´È°πÁõÆÂêçÁß∞
      expect(homepageResult.content).toContain('TestProject')
      expect(aboutResult.content).toContain('TestProject')
      expect(blogResult.content).toContain('TestProject')

      // È™åËØÅÂÜÖÂÆπÈïøÂ∫¶ÂêàÁêÜ
      expect(homepageResult.content.length).toBeGreaterThan(100)
      expect(aboutResult.content.length).toBeGreaterThan(100)
      expect(blogResult.content.length).toBeGreaterThan(100)

      // È™åËØÅÁâπÂÆöÂÜÖÂÆπÁâπÂæÅ
      expect(homepageResult.content).toContain('Welcome')
      expect(aboutResult.content).toContain('About')
    })
  })

  describe('ÊèêÁ§∫È™åËØÅ', () => {
    it('Â∫îËØ•È™åËØÅÊèêÁ§∫Ë¥®Èáè', () => {
      const goodPrompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)
      const validation = promptLibrary.validatePrompt(goodPrompt, 'homepage')

      expect(validation.isValid).toBe(true)
      expect(validation.errors).toHaveLength(0)
      expect(validation.score).toBeGreaterThan(80)
    })

    it('Â∫îËØ•Ê£ÄÊµãÁ©∫ÊèêÁ§∫', () => {
      const validation = promptLibrary.validatePrompt('', 'homepage')

      expect(validation.isValid).toBe(false)
      expect(validation.errors).toContain('Prompt cannot be empty')
      expect(validation.score).toBeLessThan(100)
    })

    it('Â∫îËØ•Ê£ÄÊµãËøáÁü≠ÁöÑÊèêÁ§∫', () => {
      const shortPrompt = 'Short'
      const validation = promptLibrary.validatePrompt(shortPrompt, 'homepage')

      expect(validation.isValid).toBe(false)
      expect(validation.errors.some(error => error.includes('too short'))).toBe(true)
      expect(validation.score).toBeLessThan(100)
    })

    it('Â∫îËØ•Ê£ÄÊµãËøáÈïøÁöÑÊèêÁ§∫', () => {
      const longPrompt = 'A'.repeat(15000)
      const validation = promptLibrary.validatePrompt(longPrompt, 'homepage')

      expect(validation.isValid).toBe(false)
      expect(validation.errors.some(error => error.includes('too long'))).toBe(true)
      expect(validation.score).toBeLessThan(100)
    })

    it('Â∫îËØ•È™åËØÅÊ®°ÊùøÁâπÂÆöËßÑÂàô', () => {
      const promptWithoutFeatures = 'Create a homepage for TestProject. It is a great project.'
      const validation = promptLibrary.validatePrompt(promptWithoutFeatures, 'homepage')

      expect(validation.isValid).toBe(false)
      expect(validation.errors.some(error => error.includes('features'))).toBe(true)
    })
  })

  describe('Ë¥®ÈáèÊéßÂà∂', () => {
    it('Â∫îËØ•Á°Æ‰øùÊèêÁ§∫ÂåÖÂê´ÂÖ≥ÈîÆ‰ø°ÊÅØ', () => {
      const prompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)

      // Ê£ÄÊü•ÂÖ≥ÈîÆ‰ø°ÊÅØÂ≠òÂú®
      expect(prompt).toContain(mockProjectAnalysis.metadata.name)
      expect(prompt).toContain(mockProjectAnalysis.metadata.description)

      // Ê£ÄÊü•ÊäÄÊúØÊ†à‰ø°ÊÅØ
      const hasTechStack = mockProjectAnalysis.techStack.some(tech =>
        prompt.includes(tech)
      )
      expect(hasTechStack).toBe(true)

      // Ê£ÄÊü•ÁâπÊÄß‰ø°ÊÅØ
      const hasFeatures = mockProjectAnalysis.features.some(feature =>
        prompt.includes(feature) || prompt.toLowerCase().includes(feature.toLowerCase())
      )
      expect(hasFeatures).toBe(true)
    })

    it('Â∫îËØ•ÁîüÊàê‰∏ÄËá¥ÁöÑÊèêÁ§∫ÁªìÊûÑ', () => {
      const prompt1 = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)
      const prompt2 = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)

      // Áõ∏ÂêåËæìÂÖ•Â∫îËØ•‰∫ßÁîüÁõ∏ÂêåËæìÂá∫
      expect(prompt1).toBe(prompt2)

      // ‰ΩÜ‰∏çÂêåÊ®°ÊùøÂ∫îËØ•‰∫ßÁîü‰∏çÂêåÁªìÊûÑ
      const aboutPrompt = promptLibrary.generatePrompt('about', mockProjectAnalysis, testOptions)
      expect(prompt1).not.toBe(aboutPrompt)
    })

    it('Â∫îËØ•Â§ÑÁêÜÁâπÊÆäÂ≠óÁ¨¶ÂíåÊ†ºÂºè', () => {
      const specialAnalysis = {
        ...mockProjectAnalysis,
        metadata: {
          ...mockProjectAnalysis.metadata,
          name: 'Test-Project_2024',
          description: 'A "revolutionary" AI-powered platform with 100% accuracy!'
        }
      }

      const prompt = promptLibrary.generatePrompt('homepage', specialAnalysis, testOptions)

      expect(prompt).toContain('Test-Project_2024')
      expect(prompt).toContain('"revolutionary"')
      expect(prompt).toContain('100%')
    })
  })

  describe('ÊÄßËÉΩÊµãËØï', () => {
    it('Â∫îËØ•Âú®ÂêàÁêÜÊó∂Èó¥ÂÜÖÁîüÊàêÊèêÁ§∫', () => {
      const startTime = Date.now()

      const prompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)

      const duration = Date.now() - startTime
      expect(duration).toBeLessThan(100) // Â∫îËØ•Âú®100msÂÜÖÂÆåÊàê
      expect(prompt).toBeTruthy()
    })

    it('Â∫îËØ•ÊîØÊåÅÊâπÈáèÊèêÁ§∫ÁîüÊàê', () => {
      const templates = ['homepage', 'about', 'blog']
      const startTime = Date.now()

      const prompts = templates.map(template =>
        promptLibrary.generatePrompt(template, mockProjectAnalysis, testOptions)
      )

      const duration = Date.now() - startTime
      expect(duration).toBeLessThan(500) // ÊâπÈáèÁîüÊàêÂ∫îËØ•Âú®500msÂÜÖÂÆåÊàê
      expect(prompts).toHaveLength(3)
      prompts.forEach(prompt => expect(prompt).toBeTruthy())
    })

    it('Â∫îËØ•ÊîØÊåÅÂπ∂ÂèëAIÂÜÖÂÆπÁîüÊàê', async () => {
      const promises = [
        promptLibrary.generateContent('homepage', mockProjectAnalysis, testOptions),
        promptLibrary.generateContent('about', mockProjectAnalysis, testOptions),
        promptLibrary.generateContent('blog', mockProjectAnalysis, testOptions)
      ]

      const results = await Promise.all(promises)

      expect(results).toHaveLength(3)
      results.forEach(result => {
        expect(result.content).toBeTruthy()
        expect(result.tokens).toBeGreaterThan(0)
      })
    })
  })

  describe('ÈîôËØØÂ§ÑÁêÜ', () => {
    it('Â∫îËØ•Â§ÑÁêÜ‰∏çÂ≠òÂú®ÁöÑÊ®°Êùø', () => {
      expect(() => {
        promptLibrary.generatePrompt('nonexistent', mockProjectAnalysis, testOptions)
      }).toThrow('Template not found: nonexistent')
    })

    it('Â∫îËØ•Â§ÑÁêÜÁº∫Â§±ÁöÑÈ°πÁõÆÂàÜÊûêÊï∞ÊçÆ', () => {
      const incompleteAnalysis = {
        metadata: { name: 'Test' }
      } as any

      const prompt = promptLibrary.generatePrompt('homepage', incompleteAnalysis, testOptions)

      expect(prompt).toBeTruthy()
      expect(prompt).toContain('Test')
    })

    it('Â∫îËØ•Â§ÑÁêÜÊó†ÊïàÁöÑÈÄâÈ°π', () => {
      const invalidOptions = {
        tone: null,
        audience: undefined,
        optimizationStrategy: 'invalid'
      }

      const prompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, invalidOptions)

      expect(prompt).toBeTruthy()
      expect(prompt).toContain(mockProjectAnalysis.metadata.name)
    })
  })

  describe('VitestÁâπÊúâÂäüËÉΩ', () => {
    it('Â∫îËØ•ÊîØÊåÅÂø´ÁÖßÊµãËØï', () => {
      const prompt = promptLibrary.generatePrompt('homepage', mockProjectAnalysis, testOptions)

      // ÁßªÈô§Âä®ÊÄÅÂÜÖÂÆπËøõË°åÂø´ÁÖßÊµãËØï
      const snapshot = {
        templateUsed: 'homepage',
        containsProjectName: prompt.includes(mockProjectAnalysis.metadata.name),
        containsTechStack: prompt.includes('TypeScript'),
        containsFeatures: prompt.includes('AI'),
        promptLength: prompt.length,
        hasOptimization: prompt.includes('professional')
      }

      expect(snapshot).toMatchSnapshot()
    })

    it('Â∫îËØ•ÊîØÊåÅMockÈ™åËØÅ', () => {
      const mockFn = vi.fn()
      mockFn.mockReturnValue('mocked prompt')

      const result = mockFn('test')

      expect(result).toBe('mocked prompt')
      expect(mockFn).toHaveBeenCalledWith('test')
    })

    it('Â∫îËØ•ÊîØÊåÅÂºÇÊ≠•ÊµãËØï', async () => {
      const promise1 = promptLibrary.generateContent('homepage', mockProjectAnalysis, testOptions)
      const promise2 = aiService.generateContent('test prompt')

      const [result1, result2] = await Promise.all([promise1, promise2])

      expect(result1.content).toBeTruthy()
      expect(result2.content).toBeTruthy()
    })
  })
})
