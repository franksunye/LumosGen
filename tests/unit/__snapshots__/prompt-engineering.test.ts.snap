// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Prompt Engineering Unit Tests > Vitest特有功能 > 应该支持快照测试 1`] = `
{
  "containsFeatures": true,
  "containsProjectName": true,
  "containsTechStack": true,
  "hasOptimization": true,
  "promptLength": 578,
  "templateUsed": "homepage",
}
`;
